autoload/ale/events.vim

```diff
@@ -92,6 +92,11 @@ function! ale#events#FileChangedEvent(buffer) abort
     endif
 endfunction
 
+function! ale#events#FocusGained(buffer) abort
+    call setbufvar(a:buffer, 'force_changedtick', 1)
```

Comment:
We shouldn't force updates to be sent to language servers here and below. We should only send updates if the buffer has actually changed. If `b:changedtick` isn't updated because some other process updated the buffer and Vim didn't increment `b:changedtick`, we do something that causes `b:changedtick` to increment in response to the buffer being changed, maybe comparing the text before and after Vim lost focus. If the buffer hasn't been changed, no messages should be sent to the language servers.

Reply:
This is done so we can request server to recompute diagnostics. Unfortunate LSP lacks such method and I see you’ve also noted that — https://github.com/microsoft/language-server-protocol/issues/737

Reply:
I understand. I think we shouldn't request diagnostics again if the buffer hasn't changed.

Reply:
Do you think we can add such feature (requesting diagnostics via sending didChange events) configurable for specific LSP linters? In my case I'm working on `ocamlmerlin-lsp` which only works per-file — and if I run a build system — I want to re-request diagnostics from the linter no matter if buffer was changed or not (there could be new/no errors just because of the build process finished).

Reply:
Btw such behaviour can be also useful for non-LSP linters — they don't work with "push" model (I think LSP is designed around this) also, instead you ask / "pull" diagnostics from them.

Reply:
It won't work for all language servers. Languages servers don't always respond to `textDocument/didChange` with diagnostics. The Language Server Protocol needs to be changed to support explicitly requesting diagnostics for such a thing to work. The protocol doesn't offer any guarantees about what will trigger diagnostics being sent, when they will arrive, or if they will ever be sent. I think it's one of the flaws of the protocol.

`tsserver` does have an explicit command for requesting diagnostics and pretty much guarantees a response, so we could add a command to request for diagnostics for `tsserver`.

Reply:
Right, I understand that.

That's why I'm asking if we can make this configurable per-linter so it works with `ocamlmerlin-lsp` (either via `textDocument/didChange` or I can add some new type of request to query for diagnostics as a protocol extension).

Reply:
I think we should follow the protocol and only send a message for a document being changed when it has changed, and push for the protocol to be extended to allow user to request diagnostics explicitly, like you can with `tsserver`.

Reply:
Unfortunately I don't think I can wait till it's standardised in LSP — standard bodies move slowly. I prefer standards instead be informed by real world usage. I think I'll be able to patch it up on top of ALE, not optimal but doable.